### PR TITLE
New midweek nightly build schedule and weekend weekly-release

### PR DESCRIPTION
**Pull request by andrew-m-leonard**
Signed-off-by: Andrew Leonard <anleonar@redhat.com>

### CODE REVIEW

**Review by karianna — APPROVED** [review]
LGTM

File: pipelines/jobs/pipeline_job_template.groovy
Comment (M-Davies) [inline pipelines/jobs/pipeline_job_template.groovy:28]: A word of potential pessimistic warning 😅, be careful when adding if statements when fiddling around the templates.

I've been having [issues with this sort of thing recently](https://github.com/AdoptOpenJDK/openjdk-build/pull/2132#discussion_r552046743) and it's hard to notice sometimes if the parameter is being generated in the way you want it to or not.
If you find that the `releaseTypes` are not being determined correctly for each job and need help, I can give you some further info on my findings but I'm hoping you won't have the same issue I've been having 😄

**Review by M-Davies — APPROVED** [review]
Assuming this has been tested beforehand, this LGTM 👍

File: pipelines/jobs/pipeline_job_template.groovy
Comment (andrew-m-leonard) [reply pipelines/jobs/pipeline_job_template.groovy:28]: thanks for the tip, currently testing, and yes that if on releaseTypes was not working, I missunderstood what binding was. Currently testing

**Review by sxa — COMMENTED** [review]
With 8 and 11 arguably having more platforms than others (including Dragonwell for example, as well as Win32/J9/JDK8 and OpenJ9/aarch64) perhaps we should split 8 and 11 so they don't run on the same day?

I haven't delved into the logic too much, so if someone else can verify the syntax etc. that would be good. I think we should aim to merge this tomorrow morning as I'd like to determine the effects of removing three of the Windows test boxes on the timings overnight before we implement this.

**Comment by andrew-m-leonard** [comment]
@sxa we could have 8 & 11 on different days. I was trying to keep the "current" releases on the same nights
jdk8 Mon, Wed, Fri
jdk11 Tue, Thur
?
```
Mon    Tue    Wed    Thur    Fri
8       11     8      11     8     
15      16     15     16     15
        17            17        
```

**Comment by sxa** [comment]
Perhaps switch 15 and 16, on the basis that the next change will be to remove 15 and so it makes sense to have 16 and 17 on differnet schedules

```
Mon    Tue    Wed    Thur    Fri
8       11     8      11     8     
16      15     16     15     16
        17            17        
```

**Comment by andrew-m-leonard** [comment]
> Perhaps switch 15 and 16, on the basis that the next change will be to remove 15 and so it makes sense to have 16 and 17 on different schedules
> 
> ```
> Mon    Tue    Wed    Thur    Fri
> 8       11     8      11     8     
> 16      15     16     15     16
>         17            17        
> ```

yep, looks good i'll go with that, I like the symmetry!

File: pipelines/build/common/build_base_file.groovy
Comment (karianna) [inline pipelines/build/common/build_base_file.groovy:464]: Would it be simpler to test for the positive 'nightly' case?

File: pipelines/build/common/build_base_file.groovy
Comment (karianna) [inline pipelines/build/common/build_base_file.groovy:451]: extract to a constant?

**Review by karianna — CHANGES_REQUESTED** [review]

File: pipelines/jobs/pipeline_job_template.groovy
Comment (andrew-m-leonard) [reply pipelines/jobs/pipeline_job_template.groovy]: all tested good now

File: pipelines/build/common/build_base_file.groovy
Comment (andrew-m-leonard) [reply pipelines/build/common/build_base_file.groovy]: Constant added

File: pipelines/build/common/build_base_file.groovy
Comment (andrew-m-leonard) [reply pipelines/build/common/build_base_file.groovy]: the nightly project is a sub-path of the weekly-release, so it's actually easier this way.

**Comment by sxa** [comment]
Can you clarify here (for documentation purposes) what `weeklyReleaseFolder` will give us. What will be in there, and is it a full guarnatee that there are no additional significant disk space requirements as a result of this change?

**Review by sxa — CHANGES_REQUESTED** [review]
> I think we should aim to merge this tomorrow morning

On this basis of this comment from me earlier I'm going to reject it from now, and will remove it later on to avoid accidental merging of this PR today :-)

**Comment by andrew-m-leonard** [comment]
> Can you clarify here (for documentation purposes) what `weeklyReleaseFolder` will give us. What will be in there, and is it a full guarnatee that there are no additional significant disk space requirements as a result of this change?

So the weekly-release/ folder has triggered pipeline jobs with "Release" parameter rather than "Nightly". My intention is this folder contains no artifacts, as it uses the parent "jobs", but you might have just highlighted an oversight of mine, I will check!

**Comment by sxa** [comment]
> > Can you clarify here (for documentation purposes) what `weeklyReleaseFolder` will give us. What will be in there, and is it a full guarnatee that there are no additional significant disk space requirements as a result of this change?
> 
> So the weekly-release/ folder has triggered pipeline jobs with "Release" parameter rather than "Nightly". My intention is this folder contains no artifacts, as it uses the parent "jobs", but you might have just highlighted an oversight of mine, I will check!

So duplicates of the just the top level `openjdkXX-pipeline` jobs with alternate default parameters? Would it be possible as an alternative to have an extra top level jobs that invokes the existing top level pipeline jobs with the alternate parameters, or are they sufficiently different that that isn't an option?

**Comment by andrew-m-leonard** [comment]
> > > Can you clarify here (for documentation purposes) what `weeklyReleaseFolder` will give us. What will be in there, and is it a full guarnatee that there are no additional significant disk space requirements as a result of this change?
> > 
> > 
> > So the weekly-release/ folder has triggered pipeline jobs with "Release" parameter rather than "Nightly". My intention is this folder contains no artifacts, as it uses the parent "jobs", but you might have just highlighted an oversight of mine, I will check!
> 
> So duplicates of the just the top level `openjdkXX-pipeline` jobs with alternate default parameters? Would it be possible as an alternative to have an extra top level jobs that invokes the existing top level pipeline jobs with the alternate parameters, or are they sufficiently different that that isn't an option?

So you're right, it had slipped my mind that the pipeline job copies all the build job artifacts...!
So yes, creating a new pipeline script that simply invokes the current job with different parameters would work, a bit more involved but I think it's the only way to avoid the extra artifacts.
thanks

**Review by karianna — APPROVED** [review]

**Comment by andrew-m-leonard** [comment]
@sxa ok i've re-worked it with a weekly-openjdkNN-pipeline that is simply a groovy script that invokes the existing pipelines with releaseType="Release", see my test in build-scripts-testing: https://ci.adoptopenjdk.net/job/build-scripts-testing/

**Comment by sxa** [comment]
Looks ok on a (very) quick scan but I'm too knackered to do a proper review this evening. I suggest @M-Davies and @karianna re-review given that the functionality has now been altered since they reviewed.

**Review by M-Davies — APPROVED** [review]
Again, assuming this has been tested, I'm still ok with this

**Comment by andrew-m-leonard** [comment]
> Again, assuming this has been tested, I'm still ok with this

Yes, fully tested here: https://ci.adoptopenjdk.net/job/build-scripts-testing/

**Comment by andrew-m-leonard** [comment]
@karianna if you're able to re-review the latest change, thank you

**Comment by karianna** [comment]
@sxa just your review to go :-)

**Review by sxa — APPROVED** [review]
I'll echo Morgan's comments and say I'm ok as long as we run some tests it once it's merged :-)

**Comment by andrew-m-leonard** [comment]
Running a jdk11 nightly xLinux test: https://ci.adoptopenjdk.net/job/build-scripts/job/openjdk11-pipeline/1342/

**Pull request merged**
